pkg/multicluster/watcher.go

```diff
@@ -0,0 +1,216 @@
+package multicluster
+
+import (
+	"context"
+	"sync"
+
+	"github.com/rotisserie/eris"
+	"github.com/solo-io/go-utils/contextutils"
+	"github.com/solo-io/skv2/pkg/generated/kubernetes/core/v1/controller"
+	"github.com/solo-io/skv2/pkg/multicluster/kubeconfig"
+	"github.com/solo-io/skv2/pkg/reconcile"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	// MasterCluster is the clusterName for the cluster ClusterWatcher watches.
+	MasterCluster = ""
+)
+
+// ClusterWatcher watches for KubeConfig secrets on the master cluster.
+// It is responsible for starting cluster managers and calling ClusterHandler functions.
+type ClusterWatcher interface {
+	// Run starts a watch for KubeConfig secrets on the cluster managed by the given manager.Manager.
+	// Note that Run will call Start on the given manager and run all registered ClusterHandlers.
+	Run(master manager.Manager) error
+	// RegisterClusterHandler adds a ClusterHandler to the ClusterWatcher.
+	RegisterClusterHandler(handler ClusterHandler)
+}
+
+// ManagerSet maintains a manager for every cluster in the system.
+type ManagerSet interface {
+	// Cluster returns a manager for the given cluster, or an error if one does not exist.
+	Cluster(cluster string) (manager.Manager, error)
+}
```

Comment:
Do we think this is necessary, in SMH we have never needed the whole manager for a different config, just the client. Maybe we don't need to implement this interface.

Reply:
I was thinking that the multicluster client would depend on ManagerSet. something like:

```go
type Client struct {
  managers ManagerSet
}

func (c Client) Cluster(cluster string) (client.Client, error) {
  return c.managers.Cluster(string).GetClient()
}
```

Maybe we don't need that intermediary though, and the ClusterWatcher itself could implement multicluster.Client without exposing the managers themselves.

Reply:
I like the added flexibility of exposing the manager here in case use cases arise. The multicluster client will expose just the client in order to simplify consumption for the common use case.